Refactor messaging attributes and specify per-message attributes (#2957)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/messaging.md

An "intermediary" receives a message to forward it to the next receiver, which
might be another intermediary or a consumer.

### Destinations and sources

> A destination is usually identified by some name unique within the messaging system instance, which might look like a URL or a simple one-word identifier.

Reviewer: ```suggestion
A destination is usually uniquely identified by name within the messaging system instance. Examples of a destination name would be a URL or a simple one-word identifier.
```

Reviewer: Just a nit/moving things around.

Traditional messaging, such as JMS, involves two kinds of destinations: *topic*s and *queue*s.

> A message that is sent (the send-operation is often called "*publish*" in this context) to a *topic* is broadcasted to all consumers that have *subscribed* to the topic.

Reviewer: Nit: Since the "queue" explanation is in a new line, would it make sense that the "topic" explanation follows the same? I ask because one is together with the other paragraph and the other not... one might think one is different than the other somehow?

> A message submitted to a queue is processed by a message *consumer* (usually exactly once although some message systems support a more performant at-least-once mode for messages with [idempotent][] processing).

Reviewer: ```suggestion
A message that is sent to a *queue* is processed by a message *consumer* (usually exactly once although some message systems support a more performant at-least-once mode for messages with [idempotent][] processing).
```

Nit: So both "cases" use the same wording.

> Entity messages are received from is called **source** here. Messages can be routed within one or multiple brokers, so message's *source* and *destination* could be different.

Reviewer: I read the comment from @carlosalberto and your explanation and it all makes sense. But I'm a bit confused with the first sentence:

> Entity messages are received from is called **source** here. 

In this section, `destination` is clearly explained, but I feel `source` is not. Maybe we should describe it more? For ex, using the same line as `A source is usually...`?

And then at the end, we can have this sentence, to explain why we have `destination` and `source` (to address @carlosalberto concerns others might get confused as of why)? 

> Messages can be routed within one or multiple brokers, so message's *source* and *destination* could be different.

Author: @joaopgrassi, sure, I added a better explanation for the source.

The grouping usually happens through some sort of "In-Reply-To:" meta information or an explicit *conversation ID* (sometimes called *correlation ID*).
Sometimes a conversation can span multiple message destinations (e.g. initiated via a topic, continued on a temporary one-to-one queue).

### Temporary and anonymous destinations

Some messaging systems support the concept of *temporary destination* (often only temporary queues) that are established just for a particular set of communication partners (often one to one) or conversation.

> Often such destinations are also  *anonymous* - unnamed or have an auto-generated name.

Reviewer 2: ```suggestion
Often such destinations are also unnamed (anonymous) or have an auto-generated name.
```

This can be assumed if it is statically derived from application code or configuration.
Wherever possible, the real destination names after resolving logical or aliased names SHOULD be used.
If the destination name is dynamic, such as a [conversation ID](#conversations) or a value obtained from a `Reply-To` header, it SHOULD NOT be used for the span name.
In these cases, an artificial destination name that best expresses the destination, or a generic, static fallback like `"(anonymous)"` for [anonymous destinations](#temporary-and-anonymous-destinations) SHOULD be used instead.

The values allowed for `<operation name>` are defined in the section [Operation names](#operation-names) below.
If the format above is used, the operation name MUST match the `messaging.operation` attribute defined for message consumer spans below.

Examples:

* `shop.orders publish`
* `shop.orders receive`
* `shop.orders process`
* `print_jobs publish`
* `topic with spaces process`
* `AuthenticationRequest-Conversations process`

> * `(anonymous) publish` (`(anonymous)` being a stable identifier for randomly generated, anonymous destination names)

Reviewer 2: ```suggestion
* `(anonymous) publish` (`(anonymous)` being a stable identifier for automatically generated, anonymous destination names)
```

| Value  | Description |
|---|---|
| `publish` | publish |
| `receive` | receive |
| `process` | process |
<!-- endsemconv -->

Additionally `net.peer.port` from the [network attributes][] is recommended.
Furthermore, it is strongly recommended to add the [`net.transport`][] attribute and follow its guidelines, especially for in-process queueing systems (like [Hangfire][], for example).
These attributes should be set to the broker to which the message is sent/from which it is received.

Note that attributes in `messaging.message` namespace describe an individual message, `messaging.destination` namespace
contains attributes that describe the logical entity messages are published to, and `messaging.source` describes
logical entity messages are received from. Attributes in `messaging.batch` namespace describe batch properties.

[network attributes]: span-general.md#general-network-connection-attributes
[`net.transport`]: span-general.md#network-transport-attributes
[Hangfire]: https://www.hangfire.io/

### Producer attributes

> Following additional attributes describe message producer operations.

Reviewer 3: ```suggestion
The following additional attributes describe message producer operations.
```
Besides the minor rewording suggestion, I have a comment that I don't think the destination attributes should be considered to be producer-only. For a message published to a topic, then stored on a queue for a consumer, when the consumer receives the message, the source is the queue itself, but the destination is the topic the message was published to. If all of this information is available, I think it's useful to provide it in a span since the topic provides more information about the message than just the queue it came from.

As written, the destination attributes are implied to be relevant only to producer spans.

Author: thanks, I added destination attributes to consumer section

| Value  | Description |
|---|---|
| `publish` | publish |
| `receive` | receive |
| `process` | process |
<!-- endsemconv -->

Additionally `net.peer.port` from the [network attributes][] is recommended.
Furthermore, it is strongly recommended to add the [`net.transport`][] attribute and follow its guidelines, especially for in-process queueing systems (like [Hangfire][], for example).
These attributes should be set to the broker to which the message is sent/from which it is received.

## Attribute namespaces

- `messaging.message`: Contains attributes that describe individual messages
- `messaging.destination`: Contains attributes that describe the logical entity messages are published to
- `messaging.source`: Contains attributes that describe the logical entity messages are received from
- `messaging.batch`: Contains attributes that describe batch operations
- `messaging.consumer`: Contains attributes that describe message consumers

Note: Messaging system-specific attributes MUST be defined in the corresponding `messaging.{system}` namespace
as described in [Attributes specific to certain messaging systems](#attributes-specific-to-certain-messaging-systems).

> contains attributes that describe the logical entity messages are published to, and `messaging.source` describes

Author: good catch, thank you!

[Hangfire]: https://www.hangfire.io/

### Producer attributes

The following additional attributes describe message producer operations.

<!-- semconv messaging.producer -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.destination.anonymous` | boolean | A boolean that is true if the message destination is anonymous (could be unnamed or have auto-generated name). |  | Conditionally Required: [1] |
| `messaging.destination.kind` | string | The kind of message destination | `queue` | Conditionally Required: [2] |
| `messaging.destination.name` | string | The message destination name [3] | `MyQueue`; `MyTopic` | Conditionally Required: [4] |
| `messaging.destination.template` | string | Low cardinality representation of the messaging destination name [5] | `/customers/{customerId}` | Conditionally Required: [6] |
| `messaging.destination.temporary` | boolean | A boolean that is true if the message destination is temporary and might not exist anymore after messages are processed. |  | Conditionally Required: [7] |

**[1]:** If value is `true`. When missing, the value is assumed to be `false`.

> **[2]:** If the message destination is either a `queue` or a `topic`.

Author: @ppatierno good point, we can discuss it at the messaging SIG meeting on Thursday. 

This PR though does not introduce anything new wrt `topic` and `queue` terminology and I'd prefer to address it separately.

> **[2]:** If the message destination is either a `queue` or a `topic`.

Reviewer 4: Ok. I will approve as it is for now ;-)

> Note: Consumer spans can have attributes describing source and destination. Since messages could be routed by brokers, source and destination don't always match. If original destination information is available on the consumer, consumer instrumentations SHOULD populate corresponding `messaging.destination` attributes.

### Per-message attributes

All messaging operations (`publish`, `receive`, `process`, or other not covered by this specification) can describe a batch of messages.
Attributes in `messaging.message` or `messaging.{system}.message` namespace describe individual messages. For single-message operations they SHOULD be set on corresponding span.

For batch operations per-message attributes are usually different and cannot be set on the corresponding span and MAY be set on links. See [Batch Receiving](#batch-receiving) and [Batch Processing](#batch-processing) for more information on correlation using links.

Some messaging systems such as Kafka, Azure EventGrid allow to publish a single batch of messages to different topics, in such cases attributes in `messaging.destination` and `messaging.source` MAY be

> set on links. Corresponding instrumentations MAY set source and destination attributes on the span if all messages in the batch share the same destination or source.

Reviewer 2: We shouldn't be too strict here.

Just an example: if this would be a MUST, then when receiving a batch of messages, one would need to go through all the messages and check whether all source attributes would be the same, before it's possible to decide whether to put attributes on the span or on links/events.

### Attributes specific to certain messaging systems

> All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Attributes that describe message, destination, source, consumer, or batch properties SHOULD be populated under corresponding namespace:

Reviewer 2: We should avoid MUST statements in semantic conventions. A MUST doesn't allow any exceptions, a SHOULD allows exceptions if there's a well-grounded reason.

> All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Attributes that describe message, destination, source, consumer, or batch properties SHOULD be populated under corresponding namespace:

Author: We had a long discussion on the SIG meeting regarding some attributes, for example, routing_key on RabbitMQ, where it's not clear if it should be on `destination` or on `message` namespace. Given that it could be hard to classify attributes, it's impossible to be strict here.

* `messaging.{system}.message` namespace describes individual messages

> * `messaging.{system}.destination` namespace describes destination a message (or a batch) are published to, and `messaging.{system}.source` describes source messages were received from. Combination of attributes in these namespaces should uniquely identify the entity and describe it as precisely as possible. For example, if partitioning is supported, attributes should include partition or other important and visible to users entity properties.

Reviewer: ```suggestion
* `messaging.{system}.destination` and `messaging.{system}.source`: Describes the destination and source a message (or a batch) are published to and received from respectively. The combination of attributes in these namespaces should uniquely identify the entity and describe it as precisely as possible. For example, if partitioning is supported, the attributes should include the partition or other important and visible to users entity properties.
```

I think the last sentence needs some love, but I was ensure of what the goal of it was. I left it as is we can discuss more.

Author: I gave it a bit more love, lmk if you think it's enough :)

* `messaging.{system}.consumer` namespace describes message consumer properties
* `messaging.{system}.batch` namespace describes message batch properties

> 

Reviewer 2: I like that idea. I think it would be helpful to have the description of these high-level namespaces (`messaging.source`, `messaging.destination`, `messaging.consumer`, `messaging.message`), and then mention that all of them can be nested under `messaging.{system}`.

> 

Author: I have a similar, but a non-normative sentence on general attributes. I updated it to mention that system-specific attributes must be on the corresponding namespace:

> Note that attributes in `messaging.message` namespace describe an individual message, `messaging.destination` namespace
contains attributes that describe the logical entity messages are published to, and `messaging.source` describes
logical entity messages are received from; attributes in `messaging.batch` namespace describe batch properties and `messaging.consumer` namespace
describes message consumer properties.
Messaging system-specific attributes MUST be defined in the corresponding `messaging.{system}` namespace
as described in [Attributes specific to certain messaging systems](#attributes-specific-to-certain-messaging-systems).

> 

Reviewer: I left a suggestion above. The text there is already good so feel free to follow the suggestion or not :)

> 

Author: Thank you, I like it!

set on links. Corresponding instrumentations MAY set source and destination attributes on the span if all messages in the batch share the same destination or source.

### Attributes specific to certain messaging systems

All attributes that are specific for a messaging system SHOULD be populated in `messaging.{system}` namespace. Attributes that are specific to the message, the message destination, the message source, or the whole batch SHOULD be populated under corresponding namespace:

* message-specific under `messaging.{system}.message`
* destination-specific under `messaging.{system}.destination`
* source-specific under `messaging.{system}.source`
* batch-specific under `messaging.{system}.batch`

#### RabbitMQ

In RabbitMQ, the destination is defined by an *exchange* and a *routing key*.
`messaging.destination.name` MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.

<!-- semconv messaging.rabbitmq -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|

> | `messaging.rabbitmq.message.routing_key` | string | RabbitMQ message routing key. | `myKey` | Conditionally Required: If not empty. |

Reviewer 5: Summarizing here my take on the discussion we had around this today at the SIG.

As written few lines above:
> In RabbitMQ, the destination is defined by an *exchange* and a *routing key*.

I think they are considered one unit that describes where the message is sent to and how it is routed when processed by rabbit broker, which are both a property of the "destination" (the rabbit broker that receives this message)

In rabbit [routing example](https://www.rabbitmq.com/tutorials/tutorial-four-python.html), the following code:
```py
channel.basic_publish(
    exchange='direct_logs', routing_key='info', body=message)
```

Would create the following attributes
```
messaging.destination.name: "direct_logs"
messaging.rabbitmq.message.routing_key: "info"
```

Storing the routing key as `messaging.rabbitmq.*destination*.routing_key: "info" `  makes more sense to me personally.

This is a minor issue and I am good with both options

Author: I moved routing_key to destination

<!-- semconv messaging.kafka -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `messaging.kafka.message.key` | string | Message keys in Kafka are used for grouping alike messages to ensure they're processed on the same partition. They differ from `messaging.message.id` in that they're not unique. If the key is `null`, the attribute MUST NOT be set. [1] | `myKey` | Recommended |
| `messaging.kafka.consumer_group` | string | Name of the Kafka Consumer Group that is handling the message. Only applies to consumers, not producers. | `my-group` | Recommended |
| `messaging.kafka.client_id` | string | Client Id for the Consumer or Producer that is handling the message. | `client-5` | Recommended |

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 2: Another point that occurred to me after the discussion is related to the granularity of how a distinct destination is defined. In my understanding, any different combination of `source` or `destination` attributes (name, kind, anonymous, temporary, template) identifies a different destination or source. However, if there's are destinations/sources with the same name and kind, but different partitions, I think one wouldn't see this as a different destination.

The main question then is, whether the definition of a destination should go beyond the granularity of a topic or queue, or if this additional granularity should be handled at the level of a message.

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 5: > Another point that occurred to me after the discussion is related to the granularity of how a distinct destination is defined. In my understanding, any different combination of `source` or `destination` attributes (name, kind, anonymous, temporary, template) identifies a different destination or source. However, if there's are destinations/sources with the same name and kind, but different partitions, I think one wouldn't see this as a different destination.
> 
> The main question then is, whether the definition of a destination should go beyond the granularity of a topic or queue, or if this additional granularity should be handled at the level of a message.

Interesting observation. It makes sense to me.
Maybe the criteria we are looking for to host something in the `destination`/`source` namespace is whether it affects the **logical** routing in the messaging system routing model? In contrast to `partition` which is implementation / technical detail?

In this criteria, I think the `messaging.kafka.consumer_group` attribute should belong to the `source` namespace

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 4: > However, if there's are destinations/sources with the same name and kind, but different partitions, I think one wouldn't see this as a different destination.

I am not getting this sentence if I think in terms of Kafka. You cannot have two topics with the same name and different number of partitions. I think I haven't got what you mean here.

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Author: > In this criteria, I think the messaging.kafka.consumer_group attribute should belong to the source namespace

Consumer group is not a `source` property - https://docs.confluent.io/platform/current/clients/consumer.html#consumer-groups. It's a pure consumer thing and the only thing that broker does wrt the group is the coordinator election. 

What do you folks think about adding `consumer` namespace? 
We also have `messaging.consumer_id` - should we move it to `messaging.consumer.id` and then the consumer group will be in `messaging.kafka.consumer.group`

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Author: > The main question then is, whether the definition of a destination should go beyond the granularity of a topic or queue, or if this additional granularity should be handled at the level of a message.

I believe it should describe the destination as precisely as possible. At the same time, backends can make sense of attributes in different ways. For example, service map probably should not show X nodes for kafka since there are X partitions, but when your monitor consumer lag or number of sent messages, you might want to see them per partition.

Do you think we need to change the definition of the destination to reflect it?

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 2: > I believe it should describe the destination as precisely as possible.

I think then the partition should go into the `destination`/`source`.

> At the same time, backends can make sense of attributes in different ways. For example, service map probably should not show X nodes for kafka since there are X partitions, but when your monitor consumer lag or number of sent messages, you might want to see them per partition.

Agreed.

> Do you think we need to change the definition of the destination to reflect it?

Yes. I think the directive of "describe the destination as precisely as possible" is vague enough, but still very useful.

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 2: > What do you folks think about adding consumer namespace?

That would make sense to me, looking at the `messaging.consumer_id` and `messaging.consumer_group` attributes.

I think we can then say that the `consumer` namespace is by definition not message-specific?

> | `messaging.kafka.message.partition` | int | Partition the message is sent to. | `2` | Recommended |

Reviewer 2: > > However, if there's are destinations/sources with the same name and kind, but different partitions, I think one wouldn't see this as a different destination.
>
> I am not getting this sentence if I think in terms of Kafka. You cannot 
> have two topics with the same name and different number of 
> partitions. I think I haven't got what you mean here.

Not at the same time, but if you increase the partition count, you can have this configuration at different points in time. But my point is, that one probably still would say that its' the same destination/source.
